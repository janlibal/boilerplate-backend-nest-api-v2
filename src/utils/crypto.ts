//import config from '../config'
//import jwt, { SignOptions, VerifyOptions } from 'jsonwebtoken'
import * as crypto from 'crypto'
import * as bcrypt from 'bcrypt'
import { randomStringGenerator } from '@nestjs/common/utils/random-string-generator.util'

async function hashPassword(password: string) {
  const salt = await bcrypt.genSalt()
  return await bcrypt.hash(password, salt)
}

function makeHash() {
  return crypto
    .createHash('sha256')
    .update(randomStringGenerator())
    .digest('hex')
}

async function comparePasswords(
  candidatePassword: string,
  userPassword: string,
) {
  return bcrypt.compare(candidatePassword, userPassword)
}

/*const signOptions: SignOptions = {
  algorithm: config.auth.createOptions.algorithm,
  expiresIn: config.auth.createOptions.expiresIn,
  issuer: `${config.auth.createOptions.issuer}.${config.server.environment}`,
}

const verifyOptions: VerifyOptions = {
  algorithms: [config.auth.createOptions.algorithm],
  issuer: `${config.auth.createOptions.issuer}.${config.server.environment}`,
}*/

/*async function comparePasswords(
  candidatePassword: string,
  userPassword: string,
) {
  return bcrypt.compare(peperify(candidatePassword), userPassword)
}*/

/*async function hashPasswordOld(password: string) {
  return bcrypt.hash(password, saltRounds)
}/*

/*async function hashPassword(password: string) {
  return await bcrypt.hash(peperify(password), saltRounds)
}*/

/*function peperify(password: string) {
  return crypto.createHash('sha256').update(password).digest('hex')
}/*

//const hash = crypto.createHash('sha256').update(randomStringGenerator()).digest('hex')

/*async function hashPasswordOld(password: string) {
  return bcrypt.hash(peperify(password), config.auth.saltRounds)
}

function peperify(password: string) {
  return crypto
    .createHmac('sha1', config.auth.secret)
    .update(password)
    .digest('hex')
}

async function generateAccessToken(userId: string) {
  const payload = { userId }
  return jwt.sign(payload, config.auth.secret, signOptions)
}

async function verifyToken(token: string) {
  try {
    const data = await jwt.verify(token, config.auth.secret, verifyOptions)
    return data
  } catch (err) {
    if (err instanceof jwt.JsonWebTokenError || err instanceof SyntaxError) {
      return null
    }
    throw err
  }
}*/

export default {
  hashPassword,
  //hashPasswordOld,
  //generateAccessToken,
  comparePasswords,
  //peperify,
  makeHash,
  //verifyToken
}
