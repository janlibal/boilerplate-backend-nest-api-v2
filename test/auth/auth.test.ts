import { describe, beforeEach, afterEach, it, expect } from 'vitest'
import request from 'supertest'
import {
  API_PREFIX,
  APP_URL,
  TESTER_EMAIL,
  TESTER_PASSWORD,
} from '../utils/constants'
import { PrismaClient } from '@prisma/client'
import crypto from '../../src/utils/crypto'

const prisma = new PrismaClient()

describe('Auth', () => {
  const app = APP_URL
  const prefix = API_PREFIX

  beforeEach(async () => {
    await prisma.$transaction([prisma.session.deleteMany()])
    await prisma.$transaction([prisma.user.deleteMany()])

    const usr = {
      firstName: 'Joe',
      lastName: 'Doe',
      password: await crypto.hashPassword('Password123!'),
      email: 'joe.doe@joedoe.com',
    }
    await prisma.$transaction([prisma.user.create({ data: usr })])
  })

  /*describe('Logged in user', () => {
        let newUserApiToken: any
        let bareToken: any
    
        beforeAll(async () => {
          await request(app)
            .post(`${prefix}/auth/email/login`)
            .send({ email: TESTER_EMAIL, password: TESTER_PASSWORD })
            .expect(200)
            .expect(({ body }) => {
              expect(body.result.token).toMatch(
                /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/,
              )
            })
            .then(({ body }) => {
              bareToken = body.result.token
              newUserApiToken = 'jwt ' + bareToken
            })
        })
    
        it('should retrieve your own profile: /api/v1/auth/me (GET)', async () => {
          await request(app)
            .get(`${prefix}/auth/me`)
            .set('Authorization', newUserApiToken)
            .send()
            .expect(200)
            .expect(({ body }) => {
              expect(body.status).toBe(true)
              expect(body.path).toMatch('/auth/me')
              expect(body.statusCode).toBe(200)
              expect(body.timestamp).toMatch(
                /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
              )
              expect(typeof body.result.firstName).toBe('string')
              expect(typeof body.result.lastName).toBe('string')
              expect(typeof body.result.email).toBe('string') //toMatch(/^\S+@\S+\.\S+$/)
              expect(typeof body.result.provider).toBe('string')
              expect(typeof body.result.role.id).toBe('number')
              expect(typeof body.result.status.id).toBe('number')
            })
        })
    
        it('should fail with no token: /api/v1/auth/me (GET)', async () => {
          await request(app)
            .get(`${prefix}/auth/me`)
            //.set('Authorization', newUserApiToken)
            .send()
            .expect(401)
            .expect(({ body }) => {
              expect(body.status).toBe(false)
              expect(body.path).toMatch('/auth/me')
              expect(body.statusCode).toBe(401)
              expect(body.timestamp).toMatch(
                /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
              )
              expect(body.result.title).toMatch('Unauthorized')
              expect(body.result.detail).toMatch('Invalid Headers')
            })
        })
    
        it('should fail with with missing jwt prefix: /api/v1/auth/me (GET)', async () => {
          await request(app)
            .get(`${prefix}/auth/me`)
            .set('Authorization', bareToken)
            .send()
            .expect(401)
            .expect(({ body }) => {
              expect(body.status).toBe(false)
              expect(body.path).toMatch('/auth/me')
              expect(body.statusCode).toBe(401)
              expect(body.timestamp).toMatch(
                /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
              )
              expect(body.result.title).toMatch('Unauthorized')
              expect(body.result.detail).toMatch('No jwt')
            })
        })
    
        it('should fail with wrong token: /api/v1/auth/me (GET)', async () => {
          await request(app)
            .get(`${prefix}/auth/me`)
            .set('Authorization', 'jwt wrongToken')
            .send()
            .expect(401)
            .expect(({ body }) => {
              expect(body.status).toBe(false)
              expect(body.path).toMatch('/auth/me')
              expect(body.statusCode).toBe(401)
              expect(body.timestamp).toMatch(
                /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
              )
              expect(body.result.title).toMatch('Unauthorized')
              expect(body.result.detail).toMatch('Wrong token')
            })
        })
    
        it('should logout user: /api/v1/auth/logout (POST)', async () => {
          await request(app)
            .post(`${prefix}/auth/logout`)
            .set('Authorization', newUserApiToken)
            .send()
            .expect(204)
        })
      })*/

  describe('Login', () => {
    /*it('should successfully return data for logged user: /api/v1/auth/email/login (POST)', async () => {
            return await request(app)
              .post(`${prefix}/auth/email/login`)
              .send({ email: TESTER_EMAIL, password: TESTER_PASSWORD })
              .expect(200)
              .expect(({ body }) => {
                expect(body.status).toBe(true)
                expect(body.path).toMatch('/auth/email/login')
                expect(body.timestamp).toMatch(
                  /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
                )
                expect(body.statusCode).toBe(200)
                expect(body.result.token).toMatch(
                  /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/,
                )
                expect(body.result.refreshToken).toMatch(
                  /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/,
                )
                expect(body.result.tokenExpires).toBeDefined()
                expect(body.result.user.email).toMatch(/^\S+@\S+\.\S+$/)
                expect(body.result.user.hash).not.toBeDefined()
                expect(body.result.user.password).not.toBeDefined()
                expect(body.result.user.status.id).toBeDefined()
                expect(body.result.user.role.id).toBeDefined()
                expect(body.result.user.id).not.toBeDefined()
              })
          })*/
    it('should fail with missing email: /api/v1/auth/email/login (POST)', async () => {
      return await request(app)
        .post(`${prefix}/auth/email/login`)
        .send({
          //email: TESTER_EMAIL,
          password: TESTER_PASSWORD,
        })
        .expect(400)
        .expect(({ body }) => {
          expect(body.status).toBe(false)
          expect(body.path).toMatch('/auth/email/login')
          expect(body.statusCode).toBe(400)
          expect(body.timestamp).toMatch(
            /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
          )
          expect(body.result.title).toMatch('Bad Request')
          expect(body.result.detail).toMatch('Something went wrong')
          expect(body.result.errors[0]).toMatchSnapshot({
            message: 'Email must be in proper format',
          })
          expect(body.result.errors[1]).toMatchSnapshot({
            message: 'Email must be a string',
          })
          expect(body.result.errors[2]).toMatchSnapshot({
            message: 'Email cannot be empty',
          })
          expect(body.stack).toMatch(/BadRequestError: Bad Request Error/i)
        })
    })
    it('should fail with missing email and password: /api/v1/auth/email/login (POST)', async () => {
      return await request(app)
        .post(`${prefix}/auth/email/login`)
        .send({
          //email: TESTER_EMAIL,
          //password: TESTER_PASSWORD
        })
        .expect(400)
        .expect(({ body }) => {
          expect(body.status).toBe(false)
          expect(body.path).toMatch('/auth/email/login')
          expect(body.statusCode).toBe(400)
          expect(body.timestamp).toMatch(
            /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
          )
          expect(body.result.title).toMatch('Bad Request')
          expect(body.result.detail).toMatch('Something went wrong')
          expect(body.result.errors[0]).toMatchSnapshot({
            message: 'Email must be in proper format',
          })
          expect(body.result.errors[1]).toMatchSnapshot({
            message: 'Email must be a string',
          })
          expect(body.result.errors[2]).toMatchSnapshot({
            message: 'Email cannot be empty',
          })
          expect(body.result.errors[3]).toMatchSnapshot({
            message: 'Password cannot be empty',
          })
          expect(body.stack).toMatch(/BadRequestError: Bad Request Error/i)
        })
    })
    it('should fail with email being in wrong format: /api/v1/auth/email/login (POST)', async () => {
      return await request(app)
        .post(`${prefix}/auth/email/login`)
        .send({
          email: 'joe.doejoedoe.com', // TESTER_EMAIL,
          password: TESTER_PASSWORD,
        })
        .expect(400)
        .expect(({ body }) => {
          expect(body.status).toBe(false)
          expect(body.path).toMatch('/auth/email/login')
          expect(body.statusCode).toBe(400)
          expect(body.result.detail).toMatch('Something went wrong')
          expect(body.timestamp).toMatch(
            /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
          )
          expect(body.result.title).toMatch('Bad Request')
          expect(body.result.detail).toMatch('Something went wrong')
          expect(body.result.errors[0]).toMatchSnapshot({
            message: 'Email must be in proper format',
          })
          expect(body.stack).toMatch(/BadRequestError: Bad Request Error/i)
        })
    })
  })
})
