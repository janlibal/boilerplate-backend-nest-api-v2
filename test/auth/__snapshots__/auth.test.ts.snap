// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Auth > Login > should fail with email being in wrong format: /api/v1/auth/email/login (POST) 1`] = `
{
  "message": "Email must be in proper format",
}
`;

exports[`Auth > Login > should fail with missing email and password: /api/v1/auth/email/login (POST) 1`] = `
{
  "message": "Email must be in proper format",
}
`;

exports[`Auth > Login > should fail with missing email and password: /api/v1/auth/email/login (POST) 2`] = `
{
  "message": "Email must be a string",
}
`;

exports[`Auth > Login > should fail with missing email and password: /api/v1/auth/email/login (POST) 3`] = `
{
  "message": "Email cannot be empty",
}
`;

exports[`Auth > Login > should fail with missing email and password: /api/v1/auth/email/login (POST) 4`] = `
{
  "message": "Password cannot be empty",
}
`;

exports[`Auth > Login > should fail with missing email: /api/v1/auth/email/login (POST) 1`] = `
{
  "message": "Email must be in proper format",
}
`;

exports[`Auth > Login > should fail with missing email: /api/v1/auth/email/login (POST) 2`] = `
{
  "message": "Email must be a string",
}
`;

exports[`Auth > Login > should fail with missing email: /api/v1/auth/email/login (POST) 3`] = `
{
  "message": "Email cannot be empty",
}
`;
